### PR TITLE
Add a private whiteboard for reviewers

### LINKED ISSUE #7065: Add a private whiteboard for reviewers
For reviewers it would be great if we had a private whiteboard that we can use to write down information relevant to the add-on. The consequence if this is not added is that hard-to-determine edge cases (e.g. when use of eval is acceptable) may be overlooked, causing grief later on when the developer gets a rejection for something that has already been discussed. With the private whiteboard we have a chance to write this down without having it lost in the comments section of a previous version.

For testing we need to make sure the private whiteboard stays private to reviewers, the public whiteboard remains accessible to developers and reviewers. Upgrading should also retain public whiteboard data since data is migrated.

### PR DESCRIPTION
**Pull request by kewisch**
For reviewers it would be great if we had a private whiteboard that we can use to write down information relevant to the add-on. The consequence if this is not added is that hard-to-determine edge cases (e.g. when use of eval is acceptable) may be overlooked, causing grief later on when the developer gets a rejection for something that has already been discussed. With the private whiteboard we have a chance to write this down without having it lost in the comments section of a previous version.

This patch adds a such field, looking forward to your comments. I am not sure if I need to add a migration for the database, or if django does that automatically for new fields.

I added a background color so it is visually easy to determine what whiteboard you are writing in, similar to how security bugs have a different background. This will help avoid embarrassing moments when accidentally writing something into the developer whiteboard.

cc @wagnerand for wording and general comments.

image

 (fixes #7065)

### CODE REVIEW

**Comment by diox** [comment]
Since this is reserved for reviewers, I'd prefer to see this in a completely separate model with a nullable `OneToOne` relation to `Addon`. No need to overload `Addon` with yet another new field that we don't need most of the time.

Also, you do need a migration - If we were using django migrations it could generate it for you, but we don't so you have to add it manually to `src/olympia/migrations/`, picking the next available migration number.

**Comment by kewisch** [comment]
Thanks for the comments, I'll move this to a separate model and add the migration

**Comment by kewisch** [comment]
Hey @diox, I've made some changes as requested. As I'm not an expert in django models I'll need some feedback to make sure it is the right approach. I'm particularly concerned about some of the indirection I used to populate the `PrivateWhiteboard` model, I'm sure there is a better way. Also, the `private_whiteboard` function in `src/olympia/reviewers/views.py` could likely be done more efficiently.

I've removed some of the tests for now, will add/rewrite them once the remaining code is correct.

File: src/olympia/migrations/984-add-private-whiteboard.sql [hunk 1]
```diff
@@ -0,0 +1,9 @@
+
+CREATE TABLE `private_whiteboard` (
+    `created` datetime(6) NOT NULL,
+    `modified` datetime(6) NOT NULL,
+    `addon_id` integer NOT NULL PRIMARY KEY,
+    `content` longtext NOT NULL
+) ENGINE=InnoDB CHARACTER SET utf8 COLLATE utf8_general_ci;
+
+ALTER TABLE `private_whiteboard` ADD CONSTRAINT `addon_id_refs_id_deadbeef` FOREIGN KEY (`addon_id`) REFERENCES `addons` (`id`) ON DELETE CASCADE;
```
Comment (diox) [inline src/olympia/migrations/984-add-private-whiteboard.sql]: How did you generate this? Django typically does not generate `ON DELETE CASCADE` statements

File: src/olympia/reviewers/views.py [hunk 2]
```diff
@@ -804,6 +807,9 @@ def review(request, addon, channel=None):
     else:
         flags = []
 
+    private_whiteboard = PrivateWhiteboard.objects.get(pk=addon.pk)
+    private_wb_form = forms.PrivateWhiteboardForm(instance=private_whiteboard)
```
Comment (diox) [inline src/olympia/reviewers/views.py]: Don't abbreviate the variable, `private_whiteboard_form` + a line break is fine.

File: src/olympia/reviewers/views.py [hunk 3]
```diff
@@ -979,6 +987,20 @@ def whiteboard(request, addon, channel):
     raise PermissionDenied
 
 
+@addons_reviewer_required
```
Comment (diox) [inline src/olympia/reviewers/views.py]: Note: this will need to be changed when #6881 is fixed

File: src/olympia/migrations/984-add-private-whiteboard.sql (hunk 1 above)
Comment (kewisch) [reply src/olympia/migrations/984-add-private-whiteboard.sql]: By hand. How do I get django to do it?

Comment (diox) [reply src/olympia/migrations/984-add-private-whiteboard.sql]: `./manage.py sqlall reviewers` should do it (pick the table you added + the constraints/indexes)

File: src/olympia/reviewers/views.py (hunk 2 above)
Comment (kewisch) [reply src/olympia/reviewers/views.py]: Ok, will do. It looked super cludgy with a linebreak, and I think further down in `context()` this would cause me to wrap as well. I'll double check though.

**Comment by kewisch** [comment]
@diox this is ready to review now, I've taken care of your previous review comments.

File: src/olympia/migrations/990-add-private-whiteboard.sql [hunk 4]
```diff
@@ -0,0 +1,7 @@
+CREATE TABLE `private_whiteboard` (
+    `created` datetime(6) NOT NULL,
+    `modified` datetime(6) NOT NULL,
+    `addon_id` integer NOT NULL PRIMARY KEY,
```
Comment (diox) [inline src/olympia/migrations/990-add-private-whiteboard.sql]: This is not your fault, you did everything right, but this will fail on dev/stage/prod, because they unfortunately have a different db schema for most of the `ids`: they are `unsigned`. So the FK constraint will complain.

Can you add a comment at the top like this:
`-- Note: if the migration fails for you locally, remove the 'UNSIGNED' next to addon_id below.`
And modify `addon_id` to be `UNSIGNED`.

File: src/olympia/reviewers/models.py [hunk 5]
```diff
@@ -1170,3 +1170,15 @@ class ThemeLock(ModelBase):
 
     class Meta:
         db_table = 'theme_locks'
+
+
+class PrivateWhiteboard(ModelBase):
```
Comment (diox) [inline src/olympia/reviewers/models.py]: This would have been a nice occasion to also move the existing `whiteboard` content from `Addon` to this new model you're adding - in which case the name doesn't match. That would have avoided some extra copy/paste (you could have re-used the existing whiteboard views, setting them both at once)... but out of scope for this, so ignore this comment.

**Comment by kewisch** [comment]
The latest update moves the public whiteboard out of the `Addons` model. There might be a few rough edges, I hope this is reasonably close to what you'd expect. Sorry for the form with empty fields, I didn't see a nice way to fit that into `devhub/views.py` with the new form without refactoring so I thought I'd ask you first if you have a good idea. Happy to make further changes.

**Comment by kewisch** [comment]
@diox tests are green, this is ready for a new round of reviews.

File: src/olympia/devhub/views.py [hunk 6]
```diff
@@ -882,6 +891,15 @@ def addons_section(request, addon_id, addon, section, editable=False):
...
             prefix='dependencies')
 
+    if section == 'technical':
+        try:
+            whiteboard = Whiteboard.objects.get(pk=addon.pk)
+        except Whiteboard.DoesNotExist:
+            whiteboard = Whiteboard(pk=addon.pk)
+
+        whiteboard_form = PublicWhiteboardForm(request.POST or None,
+                                               instance=whiteboard)
```
Comment (diox) [inline src/olympia/devhub/views.py]: Add a [`prefix="whiteboard"`](https://docs.djangoproject.com/en/1.11/ref/forms/api/#prefixes-for-forms) to avoid just calling the field name `public`, cause it's a little too generic and susceptible to clashes. (this will break tests above)

File: src/olympia/reviewers/views.py [hunk 7]
```diff
@@ -1003,7 +1008,7 @@ def review(request, addon, channel=None):
         unlisted=(channel == amo.RELEASE_CHANNEL_UNLISTED),
         user_changes=user_changes_log, user_ratings=user_ratings,
         version=version, was_auto_approved=was_auto_approved,
-        whiteboard_form=forms.WhiteboardForm(instance=addon),
+        whiteboard_form=forms.WhiteboardForm(instance=whiteboard),
```
Comment (diox) [inline src/olympia/reviewers/views.py]: Also add `prefix` here

File: src/olympia/reviewers/views.py [hunk 8]
```diff
@@ -1157,10 +1162,15 @@ def whiteboard(request, addon, channel):
     perform_review_permission_checks(
         request, addon, channel, content_review_only=content_review_only)
 
-    form = forms.WhiteboardForm(request.POST or None, instance=addon)
+    whiteboard, _ = Whiteboard.objects.get_or_create(pk=addon.pk)
+    form = forms.WhiteboardForm(request.POST or None, instance=whiteboard)
```
Comment (diox) [inline src/olympia/reviewers/views.py]: And `prefix` here as well

**Comment by diox** [comment]
Also a minor conflict to fix :)

**Comment by diox** [comment]
Forgot to ask: please create an issue (and reference it in your PR description) so that QA can verify this.

**Comment by diox** [comment]
Also, migration 991 is now taken (sorry).

**Comment by kewisch** [comment]
I fixed the comments and the tests pass. mat is on PTO, maybe @EnTeQuAk can take a look at the final changes?

**Comment by EnTeQuAk** [comment]
I'll have a look later today

File: src/olympia/devhub/views.py [hunk 9]
```diff
@@ -307,9 +309,15 @@ def feed(request, addon_id=None):
 
 @dev_required
 def edit(request, addon_id, addon):
+    try:
+        whiteboard = Whiteboard.objects.get(pk=addon.pk)
+    except Whiteboard.DoesNotExist:
+        whiteboard = Whiteboard(pk=addon.pk)
```
Comment (EnTeQuAk) [inline src/olympia/devhub/views.py]: This doesn't create the `Whiteboard` instance, I think you want to either do a `whiteboard.save()` afterwards or use https://docs.djangoproject.com/en/1.8/ref/models/querysets/#django.db.models.query.QuerySet.get_or_create?

File: src/olympia/devhub/views.py [hunk 10]
```diff
@@ -882,6 +891,16 @@ def addons_section(request, addon_id, addon, section, editable=False):
             queryset=addon.addons_dependencies.all(), addon=addon,
             prefix='dependencies')
 
+    if section == 'technical':
+        try:
+            whiteboard = Whiteboard.objects.get(pk=addon.pk)
+        except Whiteboard.DoesNotExist:
+            whiteboard = Whiteboard(pk=addon.pk)
+
```
Comment (EnTeQuAk) [inline src/olympia/devhub/views.py]: Same here, this doesn't create the Whiteboard instance, I think you want to either do a whiteboard.save() afterwards or use https://docs.djangoproject.com/en/1.8/ref/models/querysets/#django.db.models.query.QuerySet.get_or_create?

File: src/olympia/migrations/992-add-private-whiteboard.sql [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+    `modified` datetime(6) NOT NULL,
+    `addon_id` integer UNSIGNED  NOT NULL PRIMARY KEY,
+    `private` longtext NOT NULL,
+    `public` longtext NOT NULL
+);
+ALTER TABLE `review_whiteboard` ADD CONSTRAINT `addon_id_refs_id_3aa22f51` FOREIGN KEY (`addon_id`) REFERENCES `addons` (`id`);
+
+-- Move the whiteboard from the addons table to the new one
+INSERT INTO `review_whiteboard` (`created`, `modified`, `addon_id`, `private`, `public`)
+    SELECT `created`, `modified`, `id`, '', `whiteboard` FROM `addons` WHERE `whiteboard` != ''
```
Comment (EnTeQuAk) [inline src/olympia/migrations/992-add-private-whiteboard.sql]: I'm sorry, I just merged a PR that introduced version 992 in migrations so this should now be 993 😆

File: src/olympia/reviewers/forms.py [hunk 12]
```diff
@@ -513,8 +514,21 @@ def __init__(self, *args, **kwargs):
 class WhiteboardForm(forms.ModelForm):
 
     class Meta:
-        model = Addon
-        fields = ['whiteboard']
+        model = Whiteboard
+        fields = ['private', 'public']
+        labels = {
+            'private': _("Private Whiteboard"),
```
Comment (EnTeQuAk) [inline src/olympia/reviewers/forms.py]: We're heavy users of single quotes, let's change that here please?

File: src/olympia/reviewers/forms.py [hunk 13]
```diff
@@ -513,8 +514,21 @@ def __init__(self, *args, **kwargs):
...
+            'public': _("Whiteboard")
+        }
+
+
+class PublicWhiteboardForm(forms.ModelForm):
+    class Meta:
+        model = Whiteboard
+        fields = ['public']
+        labels = {
+            'public': _("Whiteboard")
```
Comment (EnTeQuAk) [inline src/olympia/reviewers/forms.py]: Same here

File: src/olympia/reviewers/tests/test_models.py [hunk 14]
```diff
@@ -1492,3 +1493,51 @@ def test_verdict_info_prettifier(self):
...
+        whiteboard = Whiteboard(pk=addon.pk, public=whiteboard)
+        whiteboard.save()
+
+        return addon
+
+    def assertHasInfoSet(self, addon):
+        assert any([v.has_info_request for v in addon.versions.all()])
+
+    def assertHasInfoNotSet(self, addon):
+        assert all([not v.has_info_request for v in addon.versions.all()])
```
Comment (EnTeQuAk) [inline src/olympia/reviewers/tests/test_models.py]: We try to not follow unittest2 `assertCamelCase` style most of the times, but name these `assert_has_info_not_set` imho. Unless it's a generic helper in `amo.tests:TestCase`

File: src/olympia/reviewers/views.py [hunk 15]
```diff
@@ -909,6 +909,14 @@ def review(request, addon, channel=None):
     else:
         flags = []
 
+    try:
+        whiteboard = Whiteboard.objects.get(pk=addon.pk)
+    except Whiteboard.DoesNotExist:
+        whiteboard = Whiteboard(pk=addon.pk)
```
Comment (EnTeQuAk) [inline src/olympia/reviewers/views.py]: shouldn't use this `get_or_create` too?

File: src/olympia/devhub/views.py (hunk 10 above)
Comment (eviljeff) [reply src/olympia/devhub/views.py]: I think it's because he doesn't want to save the instance until the POST, and then when the whiteboard `is_valid` (line 936)

File: src/olympia/reviewers/views.py (hunk 15 above)
Comment (eviljeff) [reply src/olympia/reviewers/views.py]: same in these other places - it's not being saved until it's needed.  (I've not thought through the logic about whether that's going to work)

File: src/olympia/devhub/views.py (hunk 10 above)
Comment (EnTeQuAk) [reply src/olympia/devhub/views.py]: hmm, potentially true. As I wrote, I wasn't sure but if that's the case then my comments are invalid and can be ignored

Comment (kewisch) [reply src/olympia/devhub/views.py]: Yes, what Andrew said. I don't want to save just yet, otherwise it would create empty database rows just for viewing the page.

File: src/olympia/migrations/992-add-private-whiteboard.sql (hunk 11 above)
Comment (kewisch) [reply src/olympia/migrations/992-add-private-whiteboard.sql]: Ok, taking care :) I just hope I don't have to deal with 1000 :-D

File: src/olympia/reviewers/forms.py (hunk 12 above)
Comment (kewisch) [reply src/olympia/reviewers/forms.py]: Ah yes, surprised the linter didn't catch this.

File: src/olympia/reviewers/tests/test_models.py (hunk 14 above)
Comment (kewisch) [reply src/olympia/reviewers/tests/test_models.py]: This was copy/paste. I'll take care though.

File: src/olympia/reviewers/views.py (hunk 15 above)
Comment (kewisch) [reply src/olympia/reviewers/views.py]: I'm not sure about edge cases, but I did test this on my dev instance. I thought it didn't make sense to create during viewing, if that breaks anything please let me know :)

Comment (eviljeff) [reply src/olympia/reviewers/views.py]: Does `ModelForm` need an instance anyway?  Afaik in other places we don't create objects - the form creates them on save (you'll need to pass in the add-on ID, obviously.

**Comment by EnTeQuAk** [comment]
let's just get the money, r+ :)

File: src/olympia/reviewers/views.py (hunk 15 above)
Comment (kewisch) [reply src/olympia/reviewers/views.py]: I'd need to pass in the add-on id, and previous content. This comes close to providing an instance. If you can tell me how other code does this I'm happy to make changes, but for now I think we should just go ahead and push :)r
